internal/app/go-filecoin/node/test/api.go

```diff
@@ -68,7 +72,7 @@ func (c *Client) Address() string {
 }
 
 // Run runs a CLI command and returns its output.
-func (c *Client) Run(ctx context.Context, command ...string) *th.CmdOutput {
```

Comment:
Blocking: I think this is breaking the explicit intention here, we should discuss and perhaps clarify documentation further before landing this.

Please review the comments on `CmdOutput` and its methods. Failure and error are two different things. The output error should be set only if the command was unable to execute as a unix process (e.g. file not found, permission denied etc) - basically things that should never happen. There is no `RunError()` method that expects this to happen.

Otherwise, no error should be set but the processes exit status and stdout/err captured. RunFail is intended to mean the command executed but returned a non-zero exit status. If instead the process couldn't be executed at all, then we haven't tested what we wanted and the test should fail. Failure is never expected in the shouldFail sense.

The naming of error, failure etc makes this all a bit hard to work out from the code - please clarify names and/or comments if you can.

Reply:
Thanks for the clear explanation.

I understand the distinction between failure and error and would prefer to undo these changes if it is not too much effort.  However we will need to solve some problems.

The first problem that shows up when you try to use this code on actual commands is that `commands.Run` consistently propagates the same error that shows up on stderr during error (not fail).  For example `TestSwarmConnectPeersInvalid` propagates an error (`failed to parse multiaddr "/ip4/hello"`) to `Run` that blows up the "client execution error" require.  If you check stderr the same error string shows up there too.  

I'm now realizing this probably means we are implementing commands incorrectly and perhaps we can address this issue at the source. 

If we solve this problem the second problem we will face with the unmodified code is that we will assert that the exitCode is 0 for all `Run` invocations even those made by `RunFail`.  I believe this will make `RunFail` always fail.  This has a simple fix, we can move the exit code checks to `RunFail` and `RunSuccess` by adding a simple helper on `CmdOutput`.

Reply:
I'm planning to dig into cli.Run's error propagation.  If I can preserve stderr propagation while modifying commands.Run's propagation to be a cli.ExitError without modifying dependencies or doing anything too ugly I will fix in this PR.  If not I propose we keep what is in the PR.

Reply:
Moving exit code checks sounds ok.